core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -122,44 +193,81 @@ public void setStoreListenerSource(Cache<K, V> source) {
   }
 
   @Override
-  public void onEvent(final CacheEvent<K, V> event) {
+  public void onEvent(CacheEvent<K, V> event) {
     final EventType type = event.getType();
     LOGGER.trace("Cache Event notified for event type {}", type);
-    Map<EventListenerWrapper, Future<?>> notificationResults = 
-        new HashMap<CacheEventNotificationServiceImpl.EventListenerWrapper, Future<?>>(registeredListeners.size());
-    
-    for (final EventListenerWrapper wrapper: registeredListeners) {
-      if (!wrapper.config.fireOn().contains(type)) {
-        continue;
+    CacheEventWrapper<K, V> cacheEventWrapper = new CacheEventWrapper<K, V>(event);
+    if (!(syncListenersSet.isEmpty() && aSyncListenersSet.isEmpty())) {
+      if(eventThreadLocal.get().isEmpty()) {
+        eventThreadLocal.setOrdered(orderedListenerCount.get() != 0);
+      }
+      eventThreadLocal.get().add(cacheEventWrapper);
```

Comment:
This will throw if the listener is removed concurrently and you replace the real impl with the no-op one